refactor(protocol-engine): Refuse movements to non-open Thermocyclers

SyntaxColoring opened the pull request:
# Overview

When Protocol Engine executes a movement to labware that's in a Thermocycler, raise an error if that Thermocycler's lid isn't open. This reimplements [preexisting](https://github.com/Opentrons/opentrons/blob/27bf255b92346e86c721ea9b9da9c39fdd4d6b36/api/src/opentrons/protocol_api/instrument_context.py#L1186) [behavior](https://github.com/Opentrons/opentrons/blob/de3f7e4cc9cd685e92a9912ea296e7f8d9510aab/api/src/opentrons/protocols/geometry/module_geometry.py#L309-L325) from APIv2.

Closes part of #8911.

# Changelog

* Create a `ThermocyclerMovementFlagger` unit. It's given a Protocol Engine labware ID. Through dependencies on Protocol Engine state and the underlying hardware API, it tracks down the hardware Thermocycler that contains that labware, and raises an error if its lid isn't currently open. A new `ThermocyclerNotOpenError` is added for this.
* Call `ThermocyclerMovementFlagger` from Protocol Engine's `MovementHandler` before attempting an absolute move.
* Make a couple of small fixes to adjacent code. See the GitHub diff comments.


# Review requests

* Please scrutinize my use of `hardware_api.find_modules()`. Is my usage of it correct in this context?
* The APIv2 code checked both the destination labware *and the source labware.* I'm not doing that here—I'm just checking the destination labware. Is this okay? 

   Checking the source labware didn't make sense to me; how could the pipette currently be in a closed Thermocycler?
* I will manually test this on a real robot before merging, but I haven't done so yet. Feel free to do it yourself, if that strikes your fancy.


# Risk assessment

The new code is covered well by tests, but those tests only make sense *if I'm using `hardware_api.find_modules()` correctly.*

If I get that wrong, the risk would be:

* False positives, i.e. we accidentally block movements when we shouldn't. We'd probably notice this pretty quickly.
* False negatives, i.e. we don't block movements to closed Thermocyclers like we should. This would be hard for us to notice.

SyntaxColoring committed:
refactor: Type `Thermocycler.lid_status` more precisely.

SyntaxColoring committed:
Add new `ThermocyclerNotOpenError`.

SyntaxColoring committed:
Raise on movement attempts to a non-open Thermocycler.

SyntaxColoring committed:
Elaborate todo comment.

SyntaxColoring committed:
Fix test to correctly stub out properties.

SyntaxColoring committed:
WIP: Add test.

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/modules.py:
```diff
@@ -91,6 +91,22 @@ def get_location(self, module_id: str) -> DeckSlotLocation:
         """Get the slot location of the given module."""
         return self._state.modules_by_id[module_id].location
 
+    def get_model(self, module_id: str) -> ModuleModel:
```
Needs a test.

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/modules.py:
```diff
@@ -91,6 +91,22 @@ def get_location(self, module_id: str) -> DeckSlotLocation:
+        """Get the model name of the given module."""
+        return self.get(module_id=module_id).model
+
+    def get_serial(self, module_id: str) -> str:
```
Needs a test.

SyntaxColoring commented on api/src/opentrons/protocol_engine/execution/movement.py:
```diff
@@ -171,3 +183,61 @@ async def home(self, axes: Optional[Sequence[MotorAxis]]) -> None:
+
+    def _find_thermocycler_by_serial(self, serial_number: str) -> HardwareThermocycler:
+        for attached_module in self._hardware_api.attached_modules:
+            # Different module types have different keys under .device_info.
+            # Thermocyclers should always have .device_info["serial"].
```
I think another PR should make this explicit in the type signature of `.device_info`. For example, make it return a `TypedDict` instead of a `Dict[str, str]`. Or make it explicit in the docs that `.device_info` is unstable and undefined—in which case, we either need a better way to get the serial number, or we need to find a way to not rely on the serial number.

SyntaxColoring committed:
Fix missing module handling in `get_location()` and `get_dimensions()`.

SyntaxColoring committed:
Add tests for `get_model()` and `get_serial()`.

SyntaxColoring committed:
WIP: Parametrize `MovementHandler` test.

SyntaxColoring committed:
Split TC movement flagging into a utility class.

SyntaxColoring commented on api/src/opentrons/protocol_engine/execution/movement.py:
```diff
@@ -171,3 +183,61 @@ async def home(self, axes: Optional[Sequence[MotorAxis]]) -> None:
+            return None
+
+    def _find_thermocycler_by_serial(self, serial_number: str) -> HardwareThermocycler:
+        for attached_module in self._hardware_api.attached_modules:
```
Per discussions with @sanni-t and @sfoster1, this should use `.find_modules()` instead of `.attached_modules`.

SyntaxColoring committed:
Test TC movement flagging independently.

SyntaxColoring committed:
Fix `None` lid status not being passed along correctly in error message.

SyntaxColoring committed:
Reorganize execution/__init__.py.

SyntaxColoring committed:
refactor: Move ThermocyclerMovementFlagger to its own file.

SyntaxColoring committed:
Add missing test fixtures.

SyntaxColoring committed:
Resolve todo about what to do when the hardware TC has gone missing.

SyntaxColoring committed:
Verify `ThermocyclerMovementFlagger` calls in `MovementHandler` tests.

SyntaxColoring committed:
Add tests for cases where labware isn't on a Thermocycler.

SyntaxColoring committed:
Use `.find_modules()` instead of `.attached_modules()`.

SyntaxColoring committed:
Revert unnecessary import formatting changes.

SyntaxColoring commented on api/src/opentrons/hardware_control/modules/thermocycler.py:
```diff
@@ -364,8 +364,10 @@ def lid_temp(self) -> Optional[float]:
         )
 
+    # todo(mm, 2021-11-29): Instead of being Optional, should this return
+    # ThermocyclerLidStatus.UNKNOWN when self._listener.state is None?
     @property
-    def lid_status(self) -> Optional[str]:
+    def lid_status(self) -> Optional[ThermocyclerLidStatus]:
         return self._listener.state.lid_status if self._listener.state else None
```
`self._listener.state.lid_status` was already the `ThermocyclerLidStatus` enum. This just propagates that type up, so the calling Protocol Engine code doesn't have to work with raw strings.

It wasn't immediately clear to me what it means for `self._listener.state` to be falsey or when that can happen, so I'm leaving the `Optional` thing alone, for now.

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/modules.py:
```diff
@@ -113,7 +129,7 @@ def get_by_serial(self, serial: str) -> Optional[LoadedModule]:
 
     def get_dimensions(self, module_id: str) -> ModuleDimensions:
         """Get the specified module's dimensions."""
-        return self._state.modules_by_id[module_id].definition.dimensions
+        return self.get(module_id).definition.dimensions
```
This fixes `module_view.get_dimensions("wrong-module-id")` raising `KeyError` when it should raise `errors.ModuleDoesNotExistError`.

Strictly, I guess we should be testing `pytest.raises(ModuleDoesNotExistError)` for each of these methods. But I feel like the extra coverage we'd gain from that would be outweighed by the increased test noise. I'm fine relying on convention and code review to make sure we get this right going forward, if y'all are.

SyntaxColoring committed:
Fix out-of-sync names in tests.

SyntaxColoring committed:
liiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiint.

mcous commented on api/src/opentrons/protocol_engine/execution/thermocycler_movement_flagger.py:
```diff
@@ -0,0 +1,151 @@
+                maybe if the module disconnects between when it was loaded into
+                Protocol Engine and when this function is called?
+        """
+        labware_location = self._state_store.labware.get_location(labware_id=labware_id)
+        if isinstance(labware_location, ModuleLocation):
+            module_id = labware_location.moduleId
+            if (
+                self._state_store.modules.get_model(module_id=module_id)
+                == PEModuleModel.THERMOCYCLER_MODULE_V1
+            ):
+                thermocycler_serial = self._state_store.modules.get_serial(
+                    module_id=module_id
+                )
```
I think parts and/or the whole block could be encoded into selector(s), if that helps testing and readability. Ideas that pop into my head:

- `labware.findParentModuleId` - Returns module ID if labware is on module, else None
- `geometry.findParentModule` - Returns module data (incl. serial) if labware is on module, else None
    - Or `geometry.findParentModuleSerial` - Maybe a little too specific
    - Maybe even `geometry.findParentModule(labware_id, module_type)`

SyntaxColoring replied:
> `labware.findParentModuleId` - Returns module ID if labware is on module, else None

I've implemented this as a private method in this class. It definitely helps readability in this class for it to be its own method. Putting it in `labware` didn't make testing meaningfully easier in my opinion, and I didn't want to unnecessarily bloat `labware`.

> `geometry.findParentModule` - Returns module data (incl. serial) if labware is on module, else None

Would have been nice, but mocking this out in the tests would have been a little painful because the return type, `LoadedModule`, is "big" (it has a module definition). This pain was the reason for creating `modules.get_model(id)` and `modules.get_serial(id)` in this PR.

mcous commented on api/src/opentrons/protocol_engine/execution/thermocycler_movement_flagger.py:
```diff
@@ -0,0 +1,151 @@
+            lid_status = await self._get_parent_thermocycler_lid_status(
+                labware_id=labware_id
+            )
+        except self._HardwareThermocyclerMissingError as e:
+            raise ThermocyclerNotOpenError(
+                "Thermocycler must be open when moving to labware inside it,"
+                " but can't confirm Thermocycler's current status."
+            ) from e
```
`self._HardwareThermocyclerMissingError` feels redundant to me, is it really needed?

SyntaxColoring replied:
Is the alternative that you have in mind simply raising `ThermocyclerNotOpenError()` from the inner function where we discover the hardware Thermocycler is missing, and letting it propagate out?

If so: no, a separate exception type isn't really needed. But I prefer it for clarity. Making a go/no-go decision is the responsibility of the outer `raise_if_labware_in_non_open_thermocycler()` function. I think it's harder to follow the control flow if the helpers that it calls can butt in and make that decision themselves. In other words, I tried to keep it so that the helpers would inform, rather than dictate.

mcous replied:
Cool, no worries. I sort of have the opposite reaction in terms of clarity, but it's a personal preference thing

mcous commented on api/src/opentrons/protocol_engine/execution/thermocycler_movement_flagger.py:
```diff
@@ -0,0 +1,151 @@
+        """Return the current lid status of the Thermocycler containing the labware.
+
+        Raises:
+            _NotInAThermocyclerError: If the labware isn't contained in a Thermocycler.
+                We need to raise an exception to signal this instead of returning None
+                because None is already a possible Thermocycler lid status.
```
This seems a little (code) smelly to me. I think this method is doing too much. I would split this up into a method that answers the question "is this labware ID inside a thermocycler" (see other comment about moving this logic to an external selector) and another method that answers the question "what is the lid status of the thermocycler with this serial number?"

mcous reviewed (commented):
Organizational suggestions. I think we need someone from embedded to answer the questions in the TODOs about the behaviors of `find_modules` and lid status before merging

sanni-t commented on api/src/opentrons/protocol_engine/execution/thermocycler_movement_flagger.py:
```diff
@@ -0,0 +1,151 @@
+                # The labware is in a module, but it's not a Thermocycler.
+                raise self._NotInAThermocyclerError()
+        else:
+            # The labware isn't in any module.
```
I think these two `self._NotInAThermocyclerError()` can be combined into one. Just the outer one should suffice.

sanni-t commented on api/src/opentrons/protocol_engine/execution/thermocycler_movement_flagger.py:
```diff
@@ -0,0 +1,151 @@
+            # todo(mm, 2021-12-01): It's unclear what it means if available_modules is
+            # non-empty AND simulating_module is non-None. Is concatenating them like
+            # this correct?
+            available_modules + [simulating_module]
```
In the hardware API, `available_modules` is populated only when the API instance is NOT a simulating one, while `simulating_module` is created only when the API _is_ a simulating instance. So, they aren't going to exist together. I don't think we need to be super cautious about such an improbable case, more so if it complicates the code.

SyntaxColoring committed:
Merge branch 'edge' into refuse_movements_to_closed_thermocycler

SyntaxColoring committed:
Simplify interpretation of `find_modules()` result.

SyntaxColoring committed:
Refactor to avoid the need for a `_NotInAThermocyclerError`.

SyntaxColoring committed:
Refactor `_find_containing_thermocycler()` for readability.

Split getting the parent module's ID into its own method.

Use early returns instead of nested if-else statements.

sfoster1 reviewed (approved):
The `find_modules` usage looks right to me, although I will say that considering it's almost guaranteed we'll add new versions of thermocycler it might be a good idea to avoid hardcoding the thermocycler module and instead have a single source of truth of what counts as a thermocycler

SyntaxColoring committed:
Merge branch 'edge' for continuous integration fixes

SyntaxColoring committed:
Add a spiel about the caveats of polling lid status.

sanni-t reviewed (approved):
🧢 
Tested on a robot
